File: 프로그래머스/Level2/스킬트리.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int solution(string skill, vector<string> skill_trees) {
    int answer = 0;
    map<char, int> skillTreeOrder;
    
    for(int i=0; i<skill.length(); i++){
        skillTreeOrder[skill[i]] = i+1;
    }
    
    for(string skt:skill_trees){
        int cnt = 1;
        bool check = true;
        
        for(int i=0; i<skt.length(); i++){
            if(skillTreeOrder[skt[i]] > cnt){
                check = false;
                break;
            }
            else if(skillTreeOrder[skt[i]] == cnt){
                cnt ++;
            }
        }
        if(check) answer ++;
    }
        
    return answer;
}

File: 프로그래머스/Level2/스킬트리.js
function solution(skill, skill_trees) {
    let answer = 0;
    
    skill_trees.forEach(elem => available(skill, elem) === true && answer ++);
    
    return answer;
}

function available(skill, tree) { 
    let hashMap = {}, k = 0;
    
    for(let i=0; i<skill.length; i++)
        hashMap = {...hashMap, [skill[i]] : i};
    
    for(let j=0; j<skill.length; j++){
        for(let i of tree){
            if(hashMap[i] === k)
                k++;
            
            else if(hashMap[i] > k)
                return false
            
            else continue;
        }
    }
    
    return true;
}
